Added rudimentary check on package names in dependencies

ElectreAAS opened the pull request:
Resolves #9270.

This PR adds a simple check to all the `<pkgname>` listed in the dependencies of a `dune-project`.

I just saw #9371, and have no opinions on it, but since they look closely related, I'm linking it here too.

Pinging @Leonidas-from-XIV & @emillon for a review.

Leonidas-from-XIV commented on test/blackbox-tests/test-cases/invalid-package-version.t:
```diff
@@ -13,6 +13,10 @@ opam.
 Building doesn't raise any error about the invalid version.
```
This should be updated now.

Leonidas-from-XIV commented on test/blackbox-tests/test-cases/invalid-package-version.t:
```diff
@@ -13,6 +13,10 @@ opam.
+  underscores (_). Maybe you wrote "foo.1.2.3" instead of "(foo (= 1.2.3))".
+  [1]
 
 Even generating .opam files doesn't raise any error about the version.
```
Likewise.

Leonidas-from-XIV reviewed (changes requested):
Great to have your first PR! I've added a few comments, convenient that there are already some cram tests to test the change.

ElectreAAS commented on src/dune_rules/package.ml:
```diff
@@ -415,6 +415,22 @@ let decode_name ~version =
+  let name = Name.to_string dep.name in
+  if String.exists name ~f:(function
+       | 'a' .. 'z' | 'A' .. 'Z' | '0' .. '9' | '-' | '_' -> false
+       | _ -> true)
```
Realized a bit late that `decode_name` in the same file already does the same thing, I just needed to a bit of glue in between.

ElectreAAS commented:
Note: I only added the 'opam compatible' check in the `depends` field, but not the similarly formed `conflicts` and `depopts`. For the latter it seems to make sense, but the former?

Alizter commented:
@ElectreAAS We use `conflicts` as inputs to the solver IIRC so they should also be validated the same as `depends`.

ElectreAAS commented:
I just rebased the PR on main, sorry for the noise.

I have updated the expect part of [test/blackbox-tests/test-cases/pkg/invalid-version.t](https://github.com/ElectreAAS/dune/blob/main/test/blackbox-tests/test-cases/pkg/invalid-version.t), but not the text yet, as I'm quite confused as to what is the difference in tested behaviour in this test and [test/blackbox-tests/test-cases/invalid-package-version.t](https://github.com/ElectreAAS/dune/blob/main/test/blackbox-tests/test-cases/invalid-package-version.t).

Leonidas-from-XIV commented:
However should we care about `(depends)` if we're not generating opam files and never use the data?

Not sure how many packages specify depends without generating opam files, because without it it is fairly pointless at the moment.
